### PR TITLE
compiletest: don't register predefined `MSVC`/`NONMSVC` FileCheck prefixes

### PR DESCRIPTION
**Pull request by jieyouxu**
This was fragile as it was based on host target passed to compiletest,
but the user could cross-compile and run test for a different target
(e.g. cross from linux to msvc, but msvc won't be set on the target).
Furthermore, it was also very surprising as normally revision names
(other than `CHECK`) was accepted as FileCheck prefixes.

This partially reverts the `MSVC`/`NONMSVC` predefined FileCheck
prefix registration introduced historically for some codegen tests.

This makes some codegen tests more verbose since they now need to
explicitly introduce `MSVC`/`NONMSVC` revisions, but I think that's
less surprising, e.g.:

```rs
//@ revisions: MSVC NONMSVC
//@[MSVC] only-msvc
//@[NONMSVC] ignore-msvc
```

Note that revisions are not *only* FileCheck prefixes in
FileCheck-based test suites, as they also can be used
to conditionally apply certain compiletest directives.

r? @Zalathar (or reroll a `r/? compiletest` reviewer)

try-job: x86_64-msvc
try-job: i686-msvc
try-job: x86_64-mingw-1
try-job: i686-mingw

### CODE REVIEW

File: src/tools/compiletest/src/runtest.rs [hunk 1]
```diff
@@ -1958,23 +1958,16 @@ impl<'test> TestCx<'test> {
...
         }
 
-        // Some tests also expect either the MSVC or NONMSVC prefix to be defined.
-        let msvc_or_not = if self.config.target.contains("msvc") { "MSVC" } else { "NONMSVC" };
-        filecheck.arg("--check-prefix").arg(msvc_or_not);
-
-        // The filecheck tool normally fails if a prefix is defined but not used.
-        // However, we define several prefixes globally for all tests.
+        // The filecheck tool normally fails if a prefix is defined but not used. However, sometimes
+        // revisions are used to specify *compiletest* directives which are not FileCheck concerns.
```
Comment (Zalathar) [inline src/tools/compiletest/src/runtest.rs:1970]: Nitpick: If we're revising these comments, I think they should acknowledge that auto-registering revision names as check prefix is a bit sketchy, and that having to pass `--allow-unused-prefix` is an unfortunate side-effect of not knowing whether the test author actually wanted revision-specific check prefixes or not.

Comment (Zalathar) [reply src/tools/compiletest/src/runtest.rs:1970]: The main reason I haven't tried to fix this yet is that there are quite a few tests affected, and in tests that use both CHECK and revision-specific prefixes, there's a risk of accidentally only checking the CHECK ones and silently ignoring all the revision ones.

Comment (jieyouxu) [reply src/tools/compiletest/src/runtest.rs:1970]: True. I'll think about this a bit more.

Comment (jieyouxu) [reply src/tools/compiletest/src/runtest.rs:1970]: > The main reason I haven't tried to fix this yet is that there are quite a few tests affected, and in tests that use both CHECK and revision-specific prefixes, there's a risk of accidentally only checking the CHECK ones and silently ignoring all the revision ones.

Don't we register the revisions via

```rs
if let Some(rev) = self.revision {
    filecheck.arg("--check-prefix").arg(rev);
}
```

?

Comment (jieyouxu) [reply src/tools/compiletest/src/runtest.rs]: I added comments back + further clarified the current situation.

Comment (jieyouxu) [reply src/tools/compiletest/src/runtest.rs]: Opened #134510 to track this and backlinked from comments.

File: src/tools/compiletest/src/runtest.rs [hunk 2]
```diff
@@ -1958,23 +1958,35 @@ impl<'test> TestCx<'test> {
...
+        // **implicitly** registers a `FileCheck` prefix of the same name. But this is questionable:
+        // with this setup, there is no way to distinguish whether the test writer wants to:
+        //
+        // 1. Use revisions but ONLY for compiletest directive purposes. E.g. `FileCheck`
+        //    annotations should ONLY be using the default `CHECK` prefix.
+        // 2. Use revisions but ONLY for filecheck prefix purposes. This use case is somewhat
+        //    questionable; previously we implicitly injected `MSVC` for msvc host and `NONMSVC` for
+        //    non-msvc host, but that is problematic in cross-compile test scenarios.
+        // 3. Use revisions for BOTH. This conflates `compiletest` directives (e.g. conditional test
+        //    execution) and `FileCheck` directives, and tthet test writer can't really separate
```
Comment (Zalathar) [inline src/tools/compiletest/src/runtest.rs:1975]: Typo: `tthet`

**Comment by jieyouxu** [comment]
Trimmed the comment down and just stole ur comment instead. The nuisance can be left to the discussions in the linked issue.

**Pull request merged**